File: app/services/curso_service.py
from typing import List
from fastapi import HTTPException, status
from app.models.curso import Curso
from app.dtos.curso_dto import CursoCreateDTO, CursoUpdateDTO, CursoResponseDTO
from app.repositories.curso_repository import CursoRepository
from app.repositories.estudiante_repository import EstudianteRepository
from app.repositories.profesor_repository import ProfesorRepository

class CursoService:
    def __init__(
        self,
        curso_repo: CursoRepository,
        estudiante_repo: EstudianteRepository,
        profesor_repo: ProfesorRepository
    ):
        self.curso_repo = curso_repo
        self.estudiante_repo = estudiante_repo
        self.profesor_repo = profesor_repo

    def obtener_todos_los_cursos(self) -> List[CursoResponseDTO]:
        cursos = self.curso_repo.get_all()
        return [CursoResponseDTO.model_validate(c) for c in cursos]

    def obtener_curso_por_id(self, curso_id: int) -> CursoResponseDTO:
        curso = self.curso_repo.get_by_id(curso_id)
        if not curso:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Curso con ID {curso_id} no encontrado"
            )
        return CursoResponseDTO.model_validate(curso)

    def crear_curso(self, dto: CursoCreateDTO) -> CursoResponseDTO:
        # Validación de regla de negocio: Estudiante debe existir
        if not self.estudiante_repo.get_by_id(dto.estudiante_id):
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"No se puede crear el curso. El estudiante con ID {dto.estudiante_id} no existe."
            )
        # Validación de regla de negocio: Profesor debe existir
        if not self.profesor_repo.get_by_id(dto.profesor_id):
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"No se puede crear el curso. El profesor con ID {dto.profesor_id} no existe."
            )
            
        nuevo_curso = Curso(
            nombre=dto.nombre,
            estudiante_id=dto.estudiante_id,
            profesor_id=dto.profesor_id,
            calificacion=dto.calificacion
        )
        guardado = self.curso_repo.create(nuevo_curso)
        return CursoResponseDTO.model_validate(guardado)

    def actualizar_curso(self, curso_id: int, dto: CursoUpdateDTO) -> CursoResponseDTO:
        curso = self.curso_repo.get_by_id(curso_id)
        if not curso:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Curso con ID {curso_id} no encontrado para actualizar"
            )
            
        if dto.estudiante_id is not None:
            if not self.estudiante_repo.get_by_id(dto.estudiante_id):
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"El estudiante con ID {dto.estudiante_id} no existe"
                )
            curso.estudiante_id = dto.estudiante_id
            
        if dto.profesor_id is not None:
            if not self.profesor_repo.get_by_id(dto.profesor_id):
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"El profesor con ID {dto.profesor_id} no existe"
                )
            curso.profesor_id = dto.profesor_id
            
        if dto.nombre is not None:
            curso.nombre = dto.nombre
        if dto.calificacion is not None:
            curso.calificacion = dto.calificacion

        actualizado = self.curso_repo.update(curso)
        return CursoResponseDTO.model_validate(actualizado)

File: hecho con java/src/main/java/com/umanizales/notasya/services/CursoService.java
package com.umanizales.notasya.services;

import com.umanizales.notasya.dtos.CursoDTO;
import com.umanizales.notasya.exceptions.ResourceNotFoundException;
import com.umanizales.notasya.models.Curso;
import com.umanizales.notasya.models.Estudiante;
import com.umanizales.notasya.models.Profesor;
import com.umanizales.notasya.repositories.CursoRepository;
import com.umanizales.notasya.repositories.EstudianteRepository;
import com.umanizales.notasya.repositories.ProfesorRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class CursoService {

    private final CursoRepository cursoRepository;
    private final EstudianteRepository estudianteRepository;
    private final ProfesorRepository profesorRepository;

    public CursoService(CursoRepository cursoRepository, EstudianteRepository estudianteRepository, ProfesorRepository profesorRepository) {
        this.cursoRepository = cursoRepository;
        this.estudianteRepository = estudianteRepository;
        this.profesorRepository = profesorRepository;
    }

    @Transactional(readOnly = true)
    public List<CursoDTO.Response> obtenerTodosLosCursos() {
        return cursoRepository.findAll().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public CursoDTO.Response obtenerCursoPorId(Long id) {
        Curso c = cursoRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Curso con ID " + id + " no encontrado"));
        return mapToResponse(c);
    }

    @Transactional
    public CursoDTO.Response crearCurso(CursoDTO.Create dto) {
        Estudiante est = estudianteRepository.findById(dto.getEstudianteId())
                .orElseThrow(() -> new ResourceNotFoundException("Estudiante con ID " + dto.getEstudianteId() + " no existe"));

        Profesor prof = profesorRepository.findById(dto.getProfesorId())
                .orElseThrow(() -> new ResourceNotFoundException("Profesor con ID " + dto.getProfesorId() + " no existe"));

        Curso curso = Curso.builder()
                .nombre(dto.getNombre())
                .estudiante(est)
                .profesor(prof)
                .calificacion(dto.getCalificacion())
                .build();

        Curso guardado = cursoRepository.save(curso);
        return mapToResponse(guardado);
    }

    @Transactional
    public CursoDTO.Response actualizarCurso(Long id, CursoDTO.Update dto) {
        Curso c = cursoRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Curso con ID " + id + " no encontrado"));

        if (dto.getNombre() != null) c.setNombre(dto.getNombre());
        if (dto.getCalificacion() != null) c.setCalificacion(dto.getCalificacion());

        if (dto.getEstudianteId() != null) {
            Estudiante est = estudianteRepository.findById(dto.getEstudianteId())
                    .orElseThrow(() -> new ResourceNotFoundException("Estudiante con ID " + dto.getEstudianteId() + " no existe"));
            c.setEstudiante(est);
        }

        if (dto.getProfesorId() != null) {
            Profesor prof = profesorRepository.findById(dto.getProfesorId())
                    .orElseThrow(() -> new ResourceNotFoundException("Profesor con ID " + dto.getProfesorId() + " no existe"));
            c.setProfesor(prof);
        }

        Curso actualizado = cursoRepository.save(c);
        return mapToResponse(actualizado);
    }

    private CursoDTO.Response mapToResponse(Curso c) {
        return CursoDTO.Response.builder()
                .id(c.getId())
                .nombre(c.getNombre())
                .estudianteId(c.getEstudiante().getId())
                .estudianteNombre(c.getEstudiante().getNombre())
                .profesorId(c.getProfesor().getId())
                .profesorNombre(c.getProfesor().getNombre())
                .calificacion(c.getCalificacion())
                .build();
    }
}
